3ds2-example/backend/src/payments/payments.service.ts

```diff
@@ -0,0 +1,229 @@
+
+@Injectable()
+export class PaymentsService {
+  API_KEY: string;
+  MERCHANT_ACCOUNT: string;
+  BASE_URL: string;
+
+  constructor(
+    private httpService: HttpService,
+    private configService: ConfigService,
+  ) {
+    this.API_KEY = this.configService.get<string>('API_KEY');
+    this.BASE_URL = this.configService.get<string>('BASE_URL');
+    this.MERCHANT_ACCOUNT = this.configService.get<string>('MERCHANT_ACCOUNT');
+  }
+
+  async postForPaymentMethods(): Promise<PaymentMethodsResponse> {
+    const url = `${this.BASE_URL}/paymentMethods`;
+
+    const headers: AxiosRequestConfig['headers'] = {
+      'x-api-key': this.API_KEY,
+      'Content-Type': 'application/json',
+    };
+    const postData = {
+      merchantAccount: this.MERCHANT_ACCOUNT,
+    };
+
+    const { data } = await firstValueFrom(
+      this.httpService.post<any>(url, postData, { headers }).pipe(
+        catchError((error: AxiosError) => {
+          console.log(error);
+          throw 'An error happened!';
+        }),
+      ),
+    );
+    return data;
+  }
+
+  /**
+   * This method is just a wrapper to route to the correct flow based on the flow parameter which will be NATIVE or REDIRECT
+   * This is an implementation detail of this test app and not related to the 3DS2 flow or integraiton itself.
+   * 
+   * For both native or redirect the more information you pass to the /payments request the higher chance for a frictionless flow (no challenge)
+   * we recommend you at least pass these fields: 
+   *  countryCode: string;
+      shopperName: {
+        firstName: string;
+        lastName: string;
+      },
+      shopperIP: string;
+      shopperEmail: string;
+      Please note: 
+      it is better to omit these fields than pass static values, only populate these with real data from your shopper, otherwise it could result in higher challenge rate
+   */
+  async postForPayments({ data, url, flow }): Promise<PaymentResponse> {
+    if (flow === 'NATIVE') {
+      // example of payment request for native flow
+      return this.postForPaymentsNative({ data, url });
+    } else {
+      // example of payment request for redirect flow
+      return this.postForPaymentsRedirect({ data, url });
+    }
+  }
+
+  async postForPaymentsRedirect({ data, url }): Promise<PaymentResponse> {
+    const checkout_url = `${this.BASE_URL}/payments`;
+    const headers: AxiosRequestConfig['headers'] = {
+      'x-api-key': this.API_KEY,
+    };
+
+    const paymentRequestData: PaymentRequest = {
+      amount: {
+        currency: 'EUR',
+        value: 1000,
+      },
+      countryCode: 'NL',
+      shopperName: {
+        firstName: 'Test',
+        lastName: 'Shopper',
+      },
+      shopperIP: 'http://192.0.2.1/',
+      shopperEmail: 'testshopper@madeupemail.com',
+      channel: PaymentRequest.ChannelEnum.Web,
+      browserInfo: data.browserInfo,
+      origin: url,
+      reference: 'AUTH_INTEGRATIONS_TEST_APP',
```

Comment:
For the reference, we recommend using a randomly generated GUID

Reply:
Example: https://github.com/adyen-examples/adyen-node-online-payments/blob/main/checkout-example/index.js#L53

Reply:
✅